### ISSUE TITLE
npm.bootstrap

### ISSUE THREAD

**Issue by tarwich** [saltstack/salt#24231]
First off... **SALT IS AWESOME!** :grin: Secondly, I don't know where to post this, so if I'm in the wrong place, please redirect me. It took me a whole day to hunt this down, and I'm trying to help others as well as perhaps get a fix for this.

I tried using [npm.bootstrap](http://docs.saltstack.com/en/latest/ref/states/all/salt.states.npm.html#salt.states.npm.bootstrap) on a node project that had an indirect reference to [node-gyp](https://github.com/TooTallNate/node-gyp), but the install process hung. I couldn't recover the machine and had to `vagrant halt` it. 

Here's what broke it:

``` sh
salt-ssh
└── vagrant@127.0.0.1
    └── sudo
        └── npm install
            └── node node-gyp rebuild
                └── su vagrant # <-- node-gyp doesn't like "sudo". Technically,
                    └── g++    # it's not su vagrant, but su $SSH_USER
```

``` yaml
# roster
shopper-vagrant:
  host: 127.0.0.1
  user: vagrant
  sudo: True # <-- NPM bootstraps as root, but node-gyp bails back to vagrant
  port: 2222
  priv: ../.vagrant/machines/default/virtualbox/private_key
```

The end result was to remove `SUDO_UID` and `SUDO_USER` from `ENV` before running it, per
[https://github.com/TooTallNate/node-gyp/issues/115#issuecomment-28490552]. Unfortunately, you can't use [environ.setenv](http://docs.saltstack.com/en/latest/ref/states/all/salt.states.environ.html#salt.states.environ.setenv) for this, because the changes aren't picked up by `npm install`. I ended up using the following to fix it:

``` yaml
# project.sls
test-project:
  cmd.script:
    - name: salt://npm-hack/npm-bootstrap.sh
    - cwd : {{ project.path }}
    - user: {{ project.user }}
    - onchanges:
      - git: webfetch-shopper
```

``` sh
#!/bin/sh
# npm-bootstrap.sh

export SUDO_UID=$(id -u)
export SUDO_USER=

npm install
```

So... Is there any way we can make `npm.bootstrap` see if `user: foo` is set, and run an alternative version of npm install? If I'm in the right place, I might try writing a fix, as the script I wrote could have been as simple as:

``` sh
SUDO_UID=$(id -u) SUDO_USER="" npm install
```

**Comment by jfindlay** [comment]
@tarwich, I've updated the npm module to automatically accomplish the same effect as your script if a user is supplied to the module functions, from the npm state, for example.  Do you think this will fix this issue?

**Comment by tarwich** [comment]
@jfindlay, that is amazing! At first, I didn't specify a user, because I was already root, but then when I saw a user issue, I instinctively tried providing `user: root` to see if it would fix it. So I think your solution will be perfect.

I'm happy to try to test if if you like, but I might need some guidance as I don't know how to run salt without installing it. In the meantime, I read your patch and it looks like just what is needed.

Thank you so much. I'm super excited to see the patch!

**Comment by jfindlay** [comment]
There are various ways to install salt, from distro packages, pip, git, etc.  [Here](https://docs.saltstack.com/en/latest/topics/installation/index.html) are our installation docs.  If you want to wait for the next release, 2015.5.3, which should come out in a month, then that's fine too.

**Comment by tarwich** [comment]
Man, I'm sorry. @jfindlay I can't get this working.

I'm on OSX. I cloned jfindlay/salt, but I was having troubles figuring out what to do with it. I'm considering running bootstrap in a virtual machine using jfindlay/salt as the remote, but I ran out of time for this evening. I finally just copied the raw version of your patch and tried running npm.bootstrap, but it hung my virtual machine as before. It's interesting that [environ.setenv](http://docs.saltstack.com/en/latest/ref/states/all/salt.states.environ.html#salt.states.environ.setenv) also doesn't work for me. Don't know if that's related.

I found this, which might be the most appropriate way to test, but it's a lot to learn for me: https://docs.saltstack.com/en/latest/topics/development/hacking.html

I'll keep hacking at it, and if the new version of salt comes out before I finish, then I'll try that as well. Right now I'm just out of time for the day. Thanks again for the efforts. 

My Vagrantfile for the machine that's hanging:
https://gist.github.com/tarwich/b709f1f09c9e1200e32e
